dartagnan/src/main/java/com/dat3m/dartagnan/utils/collections/SetUtil.java

```diff
@@ -84,4 +86,125 @@ public void clear() {
 
         }
     }
+
+    // Manages a distinct union of a modifiable and an immutable set.
+    public static final class GroundedSet<T> extends AbstractSet<T> {
+
+        private final Set<? extends T> ground;
+        private final Set<T> added;
+        private final Set<Object> removed;
+
+        public GroundedSet(Set<? extends T> groundSet) {
+            this(groundSet, new HashSet<>());
+        }
+
+        public GroundedSet(Set<? extends T> groundSet, Set<T> addedSet) {
+            this(groundSet, addedSet, new HashSet<>());
+        }
+
+        public GroundedSet(Set<? extends T> groundSet, Set<T> addedSet, Set<Object> removedSet) {
+            ground = groundSet;
+            added = addedSet;
+            removed = removedSet;
+        }
```

Comment:
These constructor can violate the (class) invariants. I assume that `RelationAnalysis` will guarantee that this does not happen, but in the general case, this is problematic.